### DISCUSSION TITLE
What is the order of execution in gauge in the following scenario ?

### DISCUSSION THREAD

**Discussion by ayankumaar** [getgauge/gauge#2011] [General]
I am running a gauge command like this:

1) gauge run specs/SampleSpec.spec:6 specs/loans/loanspec.spec:6 specs/accounts/accountspec.spec:7

2) gauge run specs/accounts/accountspec.spec:7 specs/SampleSpec.spec:6 specs/loans/loanspec.spec:6 

3) gauge run specs/loans/loanspec.spec:6 specs/accounts/accountspec.spec:7 specs/SampleSpec.spec:6  

**I have noticed that it always runs in the specified order in this command. Can I assume that the order that I specify with this command will always be preserved ?**

**Comment by sriv** [comment]
See https://docs.gauge.org/faqs.html?os=linux&language=python&ide=vscode#what-is-the-order-of-execution-of-scenarios-in-a-specification

In short, the order will be preserved if:
- it is not a parallel run
- there are no table driven spec/scenarios

**Reply by ayankumaar** [reply]
@sriv thanks for your reply. The article talks about order of execution of scenarios in a specification. But, I may run scenarios from different specifications as well. In the three commands mentioned below, I am running three different scenarios from three different specifications:

1. gauge run specs/SampleSpec.spec:6 specs/loans/loanspec.spec:6 specs/accounts/accountspec.spec:7

2. gauge run specs/accounts/accountspec.spec:7 specs/SampleSpec.spec:6 specs/loans/loanspec.spec:6

3. gauge run specs/loans/loanspec.spec:6 specs/accounts/accountspec.spec:7 specs/SampleSpec.spec:6

**In this case also the order seems to be preserved. Am I correct ? I am assuming in case of parallel and data driven execution the order will not be preserved.**

**Reply by sriv** [reply]
In the examples that you mentioned, it'll execute the scenarios in the order it is specified.

However, if you specify two scenarios from the same spec file, gauge will not honour the order of scenarios within a spec.

ex: `gauge run specs/Sample1.spec:30 specs/Sample2.spec:19 specs specs/Sample1.spec:10` - will cause all mentioned scenarios from `Sample1.spec` to be grouped together and run before `Sample2.spec`'s scenarios. And the order will be `Sample1.spec:10`, `Sample1.spec:30`, `Sample2.spec:19`.

Not saying this is ideal, and we'll work on fixing this. But this is how it behaves currently. Moreover there are issues with order when tables are involved.

**Reply by ayankumaar** [reply]
@sriv thanks for your response. Is there any other way to trigger ordered execution in gauge ?

**Reply by sriv** [reply]
No, in general gauge assumes unordered execution of scenarios, and encourages scenarios to be independent. Explicitly passing a list of specs is the only way to mandate some order at the moment.
